Fix: All translations related to JS blocks are loaded even if the block is not used

dinhtungdu opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here -->

Fixes #6023

We're using `pre_load_script_translations` to load the WC Core JSON translation. This is a filter but we also print some script inside the callback, which leads to the issue detailed in #6023.

This PR fixes and refactors our approach to using the `load_script_translation_file` filter. Instead of doing complex calculations inside our filter callback, we rely on WordPress to calculate and prepare the translation file. Then we only try loading the WooCommerce Core translation when the feature plugin translation doesn't exist.

##### Additional fix

This PR also fixes the translation loading for Mini Cart inner blocks. The Mini Cart block is a special block, we only load it partially (the icon and the lazy load script) initially. The remaining scripts are only loaded when it's actually used (customers open the Mini Cart or add a product to the cart). The translations of those lazy-loaded scripts should be lazy-loaded too.

The correct behavior has been implemented in https://github.com/woocommerce/woocommerce-blocks/pull/6158/files#diff-4c8bd3d0cfbd21fce3d29b5b50c28a5172490edbdabe90867fbf4f36dd5d3cb4 but accidentally removed in https://github.com/woocommerce/woocommerce-blocks/pull/6420/files#diff-4c8bd3d0cfbd21fce3d29b5b50c28a5172490edbdabe90867fbf4f36dd5d3cb4. This PR adds that behavior back with new improvements introduced in #6420 to avoid hard-coding build files.

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

#### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [ ] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) .
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md)
- [x] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

<!-- If your change has a visual component, add a screenshot here. A "before" screenshot would also be helpful. -->

| Before | After |
| ------ | ----- |
|        |       |

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

0. Change the language to the Netherlands then update all translations.
1. Visit a page without any blocks on the front end.
2. View source (don't use the Inspector).
3. See no unnecessary translations as the screenshot in #6023 shows.
4. Create a page with the All Products block.
5. See translations work as expected in the editor and on the front end.
6. See the translation of Cart and Checkout blocks (and their inner blocks) are working as expected.
7. Check the page source of the pages containing WooCommerce Blocks, see only related translations are loaded.

**For Mini Cart block:**
1. Add Mini Cart block to Header template.
2. Visite the home page, and see the Mini Cart.
3. Open the Mini Cart, and see the translation works as expected for both Empty and Filled views.
4. View the source of the home page.
5. DON'T see the translation for Mini Cart inner blocks.
6. Open the inspector of the home page, and see the translation of Mini Cart inner blocks.

**Test WC Core translation fallback:**
1. Remove all translation files for `woo-gutenberg-products-block` text-domain in `wp-content/languages/plugins`.
2. Keep translation files for `woocommerce` text-domain.
3. See the translation still work for our blocks. Some translation can be missed due to the lag between the feature plugin and the `woocommerce-blocks` package bundled in WC Core.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [x] Feature plugin
* [ ] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

### Changelog

> Fix: Only enqueue the relevant translations script.

dinhtungdu committed:
refactor: use `load_script_translation_file` to load WC Core if needed

dinhtungdu committed:
fix: lazy-load Mini Cart translations

Aljullu commented on woocommerce-gutenberg-products-block.php:
```diff
@@ -153,90 +153,49 @@ function() {
+	 * here is enough because `load_script_translations()` will check for its
+	 * existence before loading it.
+	 */
+	return str_replace( $file, "/plugins/{$domain}-", '/plugins/woocommerce-' );
```
I might be misunderstanding this, but if `$file` is the subject, shouldn't it be the [third parameter](https://www.php.net/manual/en/function.str-replace.php#refsect1-function.str-replace-parameters)?

Besides that, I think the md5 hash will be different between WC core and WC Blocks, so that's something we will need to replace too when trying to fetch the translations from core, no?

dinhtungdu replied:
> I might be misunderstanding this, but if $file is the subject, shouldn't it be the [third parameter](https://www.php.net/manual/en/function.str-replace.php#refsect1-function.str-replace-parameters)?

It's definitely my wrong function usage 😅

dinhtungdu replied:
> Besides that, I think the md5 hash will be different between WC core and WC Blocks, so that's something we will need to replace too when trying to fetch the translations from core, no?

You're right! I reused the existing code to fix the issue you mentioned here.

Aljullu reviewed (commented):
Thanks for investigating this and working on simplifying this part of the codebase, @dinhtungdu!

I can confirm this fixes #6023, however, it introduces the issue that if a language doesn't have a WC Blocks translation, then translations from WC core are not loaded. You can test it if you change your site language to Italian or Catalan and go to Cart or Checkout pages. When in `trunk`, they are translated, but in this branch they aren't. Can you reproduce that too?

dinhtungdu committed:
fix str_replace function call

dinhtungdu commented:
@Aljullu Thanks so much for your review!

>  When in trunk, they are translated, but in this branch they aren't. Can you reproduce that too?

I can reproduce this issue. So marking this PR as a draft until that issue is fixed.

dinhtungdu committed:
load correct translation file

dinhtungdu commented:
@Aljullu I addressed your reviews, the locales don't have WC Blocks translations now load the translation from WC Core. Tested and worked for Italian. Can you please take another look at this?

dinhtungdu committed:
use variable to improve readability

Aljullu reviewed (approved):
Everything is testing well now, and code looks good. :shipit:
